src/schemas/f142/f142_rw.h

```diff
@@ -1,93 +1,182 @@
+#include "WriterTypedBase.h"
 #include <array>
-#include <flatbuffers/flatbuffers.h>
+#include <chrono>
 #include <memory>
 #include <vector>
 
-template <typename T> using uptr = std::unique_ptr<T>;
-
 namespace FileWriter {
 namespace Schemas {
 namespace f142 {
-#include "schemas/f142_logdata_generated.h"
-using FBUF = LogData;
-
-class writer_typed_base {
-public:
-  virtual ~writer_typed_base() = default;
-  virtual h5::append_ret write_impl(FBUF const *fbuf) = 0;
-};
-
-template <typename DT, typename FV>
-class writer_typed_array : public writer_typed_base {
-public:
-  writer_typed_array(hdf5::node::Group hdf_group,
-                     std::string const &source_name, hsize_t ncols,
-                     Value fb_value_type_id, CollectiveQueue *cq);
-  writer_typed_array(hdf5::node::Group, std::string const &source_name,
-                     hsize_t ncols, Value fb_value_type_id, CollectiveQueue *cq,
-                     HDFIDStore *hdf_store);
-  ~writer_typed_array() override = default;
-  h5::append_ret write_impl(FBUF const *fbuf) override;
-  uptr<h5::h5d_chunked_2d<DT>> ds;
-  Value _fb_value_type_id = Value::NONE;
-};
 
-template <typename DT, typename FV>
-class writer_typed_scalar : public writer_typed_base {
-public:
-  writer_typed_scalar(hdf5::node::Group hdf_group,
-                      std::string const &source_name, Value fb_value_type_id,
-                      CollectiveQueue *cq);
-  writer_typed_scalar(hdf5::node::Group hdf_group,
-                      std::string const &source_name, Value fb_value_type_id,
-                      CollectiveQueue *cq, HDFIDStore *hdf_store);
-  ~writer_typed_scalar() override = default;
-  h5::append_ret write_impl(FBUF const *fbuf) override;
-  uptr<h5::h5d_chunked_1d<DT>> ds;
-  Value _fb_value_type_id = Value::NONE;
-};
+/// Because of SWMR, we have to create HDF structures first, then close
+/// everything again, and then reopen those structures from the previously
+/// created file.  This enum is used to indicate which phase we are in.
+enum class CreateWriterTypedBaseMethod { CREATE, OPEN };
 
-class FlatbufferReader : public FileWriter::FlatbufferReader {
-  bool verify(Msg const &msg) const override;
-  std::string source_name(Msg const &msg) const override;
-  uint64_t timestamp(Msg const &msg) const override;
+/// The HDFWriterModule contains a list of these DatasetInfo, which represents
+/// the essential information about the datasets used in this HDFWriterModule.
+/// By having this information in a separate list we can condense the code
+/// needed for initialization.
+struct DatasetInfo {
+  /// Name of the HDF dataset
+  std::string Name;
+  /// Size of the HDF chunks in bytes
+  size_t ChunkBytes;
+  /// Size of the buffer which is used to optimize writing of small messages
+  size_t BufferSize;
+  /// Maximum size of a message which is considered for buffering
+  size_t BufferPacketMaxSize;
+  /// Helper
+  uptr<h5::h5d_chunked_1d<uint64_t>> *Ptr;
```

Comment:
It is slightly dubious to have have `Ptr` be a pointer to a `std::unique_ptr`. Although (after checking more closely) this code appears to be correct, it is dangerous. In my opinion `Ptr` should be either a `std::shared_ptr` or a raw pointer. Regardless, it should definitely be initialised to `nullptr`.

Reply:
True, I'll fix that.